### PR TITLE
Add range clauses ([NOT] BETWEEN) support

### PR DESCRIPTION
**Pull request by denisvm**
Implementation of range clauses for BETWEEN and NOT BETWEEN.

```go
sql, _, _ = db.From("test").Where(goqu.L("(a + b)").Between(10,100)).ToSql()
fmt.Println(sql)
// prints: SELECT * FROM "test" WHERE ((a + b) BETWEEN 10 AND 100)
```

PS: I've not found a clean way yet to implement the expression mapping in `mapToExpressionList` function.

### CODE REVIEW

File: default_adapter.go [hunk 1]
```diff
@@ -726,6 +733,33 @@ func (me *DefaultAdapter) BooleanExpressionSql(buf *SqlBuilder, operator Boolean
...
+	} else {
+		return NewGoquError("Range operator %+v not supported", operatorOp)
+	}
+	rhs1 := operator.Rhs1()
+	rhs2 := operator.Rhs2()
+	buf.WriteRune(space_rune)
+	if err := me.Literal(buf, rhs1); err != nil {
+		return err
+	}
+	buf.WriteString(" AND ")
```
Comment (doug-martin) [inline default_adapter.go]: I think you can use the `default_and_fragment` here

File: expressions.go [hunk 2]
```diff
@@ -1003,6 +1025,68 @@ func checkBoolExpType(op BooleanOperation, lhs Expression, rhs interface{}, inve
...
+type (
+	RangeOperation  int
+	RangeExpression interface {
+		Expression
+		//Returns the operator for the expression
+		Op() RangeOperation
+		//The left hand side of the expression (e.g. I("a")
+		Lhs() Expression
+		//The right hand side of the expression could be a primitive value, dataset, or expression
+		Rhs1() interface{}
```
Comment (doug-martin) [inline expressions.go]: So Im torn on this you could go a couple of different ways 

1. I think Rhs1 and 2 could be joined into a single slice of what would always be two values here and you could keep with the pattern on Lhs and Rhs and change the between methods to take variadic args but that could get messy.

2. Create a new Range type `I("a").Between(Range(1,2))` which could still allow you to get the single Rhs and allow you to include between and notbetween in the `goqu.Ex` code which was commented out above.

What do you think? I lean towards option 2.

Comment (denisvm) [inline expressions.go]: Yeah, option 2 looks cleaner and less error prone :D

**Comment by denisvm** [comment]
Hello @doug-martin, is there anything missing to merge this? Thanks!

File: expressions.go [hunk 3]
```diff
@@ -1003,6 +1031,72 @@ func checkBoolExpType(op BooleanOperation, lhs Expression, rhs interface{}, inve
...
+		Rhs2() interface{}
+	}
+	ranged struct {
+		lhs Expression
+		rhs1 interface{}
```
Comment (doug-martin) [inline expressions.go]: I think we can replace the rhs1, rhs2 with a single RangeVal

**Comment by doug-martin** [comment]
@denisvm sorry I didnt notice that you had pushed up your changes to address that comment. I think the only thing is merging your rhs1/2 to a single RHS since you have added the RangeVal. Comment once that is done and Ill merge this. Great work!

File: dataset_test.go [hunk 4]
```diff
@@ -748,6 +770,10 @@ func (me *datasetTest) TestLiteralExpressionMap() {
 	assert.Equal(t, buf.String(), `("a" NOT IN ('a', 'b', 'c'))`)
 	assert.NoError(t, ds.Literal(me.Truncate(buf), Ex{"a": Op{"is": nil, "eq": 10}}))
 	assert.Equal(t, buf.String(), `(("a" = 10) OR ("a" IS NULL))`)
+	assert.NoError(t, ds.Literal(me.Truncate(buf), Ex{"a": Op{"between": RangeVal{1,10}}}))
```
Comment (TechnotronicOz) [inline dataset_test.go]: `golint` gripes about using `RangeVal{1,10}` this way. I would specify the `Start` and `End` like you do in https://github.com/doug-martin/goqu/pull/25/files#diff-7101a3ea121adf8d5b0acb5a2c520ca3R1524

**Comment by denisvm** [comment]
@doug-martin I've merged both rhs1 and rhs2 values into a single RangeVal struct as you've suggested

**Comment by doug-martin** [comment]
Thanks again for the PR. I have published under v3.2

**Pull request merged**